openassessment/xblock/submission_mixin.py

```diff
@@ -203,13 +220,24 @@ def upload_url(self, data, suffix=''):
             A URL to be used to upload content associated with this submission.
 
         """
-        if "contentType" not in data:
-            return {'success': False, 'msg': self._(u"Must specify contentType.")}
+        if 'contentType' not in data or 'filename' not in data:
+            return {'success': False, 'msg': self._(u"There was an error uploading your file.")}
         content_type = data['contentType']
+        file_name = data['filename']
+        file_name_parts = file_name.split('.')
+        file_ext = file_name_parts[-1] if len(file_name_parts) > 1 else None
+
+        if self.file_upload_type == 'image' and content_type not in self.UPLOADABLE_IMAGE_MIME_TYPES:
+            return {'success': False, 'msg': self._(u"Content type must be of an allowed image type.")}
 
-        if not content_type.startswith('image/'):
-            return {'success': False, 'msg': self._(u"contentType must be an image.")}
+        if self.file_upload_type == 'file' and content_type not in self.UPLOADABLE_FILE_MIME_TYPES:
+            return {'success': False, 'msg': self._(u"Content type must be of an allowed file type.")}
```

Comment:
Similarly,  "Content type must be one of the following types: (white list)." 
If at all possible, the items in the list should have spaces after each comma.